src/main/java/org/jetbrains/bsp/bazel/server/bsp/resolvers/actiongraph/ActionGraphV2Parser.java

```diff
@@ -0,0 +1,109 @@
+package org.jetbrains.bsp.bazel.server.bsp.resolvers.actiongraph;
+
+import com.google.common.collect.Lists;
+import com.google.devtools.build.lib.analysis.AnalysisProtosV2;
+import java.util.*;
+import java.util.stream.Collectors;
+import org.jetbrains.bsp.bazel.commons.Uri;
+
+public class ActionGraphV2Parser implements ActionGraphParser {
+  private final AnalysisProtosV2.ActionGraphContainer actionGraph;
+  private final Map<Integer, AnalysisProtosV2.PathFragment> pathFragmentMap;
+
+  public ActionGraphV2Parser(AnalysisProtosV2.ActionGraphContainer actionGraph) {
+    this.actionGraph = actionGraph;
+    pathFragmentMap =
+        actionGraph.getPathFragmentsList().stream()
+            .collect(Collectors.toMap(AnalysisProtosV2.PathFragment::getId, fragment -> fragment));
+  }
+
+  @Override
+  public List<String> getInputsAsUri(String target, String execRoot) {
+    return getInputs(target, Lists.newArrayList(".jar", "js")).stream()
+        .map(exec_path -> Uri.fromExecPath(exec_path, execRoot).toString())
+        .collect(Collectors.toList());
+  }
+
+  private List<String> getInputs(String target, List<String> suffixes) {
+    return getActions(target).stream()
+        .flatMap(action -> action.getInputDepSetIdsList().stream())
+        .flatMap(
+            depset -> {
+              Queue<Integer> queue = new ArrayDeque<>();
+              queue.add(depset);
+              return expandDepsetToArtifacts(queue).stream();
+            })
+        .map(AnalysisProtosV2.Artifact::getPathFragmentId)
+        .map(pathFragmentId -> "exec-root://" + constructPath(pathFragmentId))
+        .filter(path -> suffixes.stream().anyMatch(path::endsWith))
+        .collect(Collectors.toCollection(TreeSet::new))
+        .stream()
+        .collect(Collectors.toList());
+  }
+
+  private String constructPath(final Integer pathFragmentId) {
+    int currId = pathFragmentId;
+    Stack<String> pathStack = new Stack<>();
+    while (currId != 0) {
+      AnalysisProtosV2.PathFragment fragment = pathFragmentMap.get(currId);
+      pathStack.add(fragment.getLabel());
+      currId = fragment.getParentId();
+    }
+    final StringBuilder path = new StringBuilder();
+    while (pathStack.size() != 1) path.append(pathStack.pop()).append("/");
```

Comment:
imo the logic here would be simpler if you collected the path fragments into a list, reversed the list, and then joined its contents with "/" instead of using a stack

Reply:
Sure!